### ISSUE TITLE
Routing issues related to couchbase.num_vbuckets- org.elasticsearch.action.NoShardAvailableActionException: No shard available

### ISSUE THREAD

**Issue by lexicalparadox** [couchbaselabs/couchbase-elasticsearch-connector#189]
Hi there,

My team is using this plugin to do bulk indexing into ES. However, we did not set the following configuration property in the elasticsearch.yml directly:

"couchbase.num_vbuckets"

According to the documentation, if this is not set, the property defaults 1024 on all platforms besides Mac. The documentation also mentions that this should exactly match the number of vBuckets on the CB cluster.

Can you please elaborate on this? For each ES node with the elasticsearch-transport-couchbase plugin installed, does the following property "couchbase.num_vbuckets" have to match the TOTAL number of vbuckets on each couchbase node in the CB cluster? For example, on the CB cluster, I am seeing that each server has 1024 vBuckets + 2048 replica vBuckets for a total of 3072 vBuckets per server. So does this "couchbase.num_vbuckets" need to match 3072? Or does it need to match 1024 (which is already default)? Also, does it matter how we set the number of shards/replicas in the ES settings? Right now in ES, it defaults to 5 shards and 1 replica.

I believe I did not set this parameter correctly as I am seeing logs coming from this plugin such as:

1. "no uuid in database name" 

2. ${cluster-name}-2018-08-16.log:org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [${index_name}][couchbaseCheckpoint][vbucket774UUID]: routing [null]]

I believe those logs are related to the above property mentioned.

Thanks for all the help.

**Comment by dnault** [comment]
Hi Lex,

The "couchbase.num_vbuckets" config property should match the number of vbuckets used by Couchbase Server (64 on macOS, 1024 elsewhere). The only time you need to set this property is if Couchbase and Elasticsearch are running on different operating systems, and one of them is running on macOS.

* If Couchbase is running on macOS but Elasticsearch is not, set the value to 64.
* If Elasticsearch is running on macOS but Couchbase is not, set the value to 1024.
* Otherwise, the plugin assumes Elasticsearch and Couchabase are running on the same OS, and will pick the correct value for that OS.


> does it matter how we set the number of shards/replicas in the ES settings? Right now in ES, it defaults to 5 shards and 1 replica.

No, the plugin does not care about shard/replica settings.


> "no uuid in database name"
>
> ${cluster-name}-2018-08-16.log:org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [${index_name}][couchbaseCheckpoint][vbucket774UUID]: routing [null]]

I suspect this may indicate an XDCR configuration error.  The "Remote Bucket" field in the XDCR config should match the Elasticsearch index name. This index must already exist in Elasticsearch.

(Feel free to reopen this issue or create another one if this answer has not solved your problem!)

Thanks,
David

**Comment by lexicalparadox** [comment]
Hi David,

We're running both ES and CB on a set of linux hosts. We've done the XDCR configuration correctly because we do see replications happening for more than a million or so documents. We ensure that the ES index exists before we set up the XDCR and the remote bucket does correspond to the ES index.

Based on exception I see in the logs:

`${cluster-name}-2018-08-16.log:org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [${index_name}][couchbaseCheckpoint][vbucket774UUID]: routing [null]]`

It would seem that the "couchbase.num_vbuckets" does not have the correct value. 1024, in this case doesn't seem to be the right one. The reason I am thinking that this configuration is incorrect is because for a small percentage of documents in CB, we don't see it getting replicated in ES.

If we try to update the document in CB to trigger the change on the document, we still don't see it getting replicated. My guess is that the couchbase.num_vbuckets does not correspond to the actual vbuckets from the source CB nodes. CB probably tries to route the key for the document to a corresponding shard on the ES side but can't find the "vbucket774UUID" and thus fails to replicate.

Thoughts? Should we try to tweak this parameter to see?

Thanks a ton!
Lex

**Comment by dnault** [comment]
1024 is almost certainly the correct value for Linux (it's technically possible to configure Couchbase to use a non-standard number of vbuckets, but it's not documented or supported). I do not think this is the problem.

I have some questions that might help with troubleshooting:

What version of the plugin are you using? Some metric reporting was added in Version 3.x that might be useful for discovering why some documents aren't being replicated (you'll see the metrics in the logs, logged once a minute).

Can you provide more details from the log (stack trace, context, etc.)? This will help determine which part of the plugin is encountering the "No shard available" condition.

Does the log contain any of these strings? 
* `Error indexing bulk with document IDs`
* `Error indexing document`
* `Indexing error`

What's the value of the `couchbase.ignoreFailures` config option?

As an aside, If you are a Couchbase Enterprise Edition customer it would be good to file a support ticket at this point. Of course I will do my best to help regardless, but having folks from the support team looking at the issue might lead to a speedier resolution. Just thought I should mention it as a possibility.

Speaking of possibilities, a developer preview of version 4 of the ES connector will be available at the end of August, with a GA release planned for the end of September. In addition to adding support for Elasticsearch 6, it uses DCP (XDCR v2) to pull events out of Couchbase instead of the push model used by CAPI (XDCR v1). It's possible the issue you are seeing is related to the CAPI implementation.

**Comment by lexicalparadox** [comment]
Hi David,

We are using the following connector version:
3.0.1-cypress/elasticsearch-transport-couchbase-3.0.1-cypress-es5.6.7

Unfortunately I cannot provide these logs since my team has already cleaned the logs folder so we can't show them here. If I'm recalling correctly, I did see some of the following errors as you mentioned:

```
Error indexing bulk with document IDs
Error indexing document
Indexing error
```

I believe we saw these errors under heavy load. We saw that our queue capacity for the indexing/buk thread pools were defaulting to 200 and so a lot of bulk indexing requests were getting dropped entirely. Since then, we've upped our queue capacity to a large number (like 10,000) and has seen that this has solved this problem.
 
`couchbase.ignoreFailures` is not set, but it's defaulting to false as I am seeing in our logs.

So, I actually set the `couchbase.num_vbuckets` explicitly to 1024 and it looks to have resolved this issue. I am no longer seeing issues with `NoShardAvailableActionException` and I am also seeing that all of our data is migrated successfully from CB to ES. I'll let you know if we see any further issues.

By the way, thanks for the tip on version 4 of the ES connector. Will definitely take a look!
